src/charm.py

```diff
@@ -0,0 +1,162 @@
+        self.framework.observe(kubecontrol.relation_changed, self._setup_environment)
+
+        certificates = self.on["certificates"]
+        self.framework.observe(certificates.relation_broken, self._setup_environment)
+        self.framework.observe(certificates.relation_created, self._setup_environment)
+        self.framework.observe(certificates.relation_changed, self._setup_environment)
+
+        self.framework.observe(self.on.test_action, self._on_test_action)
+        self.framework.observe(self.on.config_changed, self._setup_environment)
+
+    def _kube_control_relation_joined(self, event: EventBase):
+        self.kube_control.set_auth_request(self.unit.name)
+        return self._setup_environment(event)
+
+    def _ensure_kube_control_relation(self, event: EventBase) -> bool:
+        self.unit.status = MaintenanceStatus("Evaluating kubernetes authentication.")
+        evaluation = self.kube_control.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        if not self.kube_control.get_auth_credentials(self.unit.name):
+            self.unit.status = WaitingStatus("Waiting for kube-control: unit credentials.")
+            return False
+        self.unit.status = MaintenanceStatus("Kubernetes authentication completed.")
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/root/.kube/config", "root", self.unit.name
+        )
+        self.kube_control.create_kubeconfig(
+            self.CA_CERT_PATH, "/home/ubuntu/.kube/config", "ubuntu", self.unit.name
+        )
+        return True
+
+    def _ensure_certificates_relation(self, event: EventBase) -> bool:
+        self.unit.status = MaintenanceStatus("Evaluating certificates.")
+        evaluation = self.certificates.evaluate_relation(event)
+        if evaluation:
+            if "Waiting" in evaluation:
+                self.unit.status = WaitingStatus(evaluation)
+            else:
+                self.unit.status = BlockedStatus(evaluation)
+            return False
+        self.CA_CERT_PATH.write_text(self.certificates.ca)
+        return True
+
+    def _setup_environment(self, event: EventBase) -> None:
+        if not self._ensure_certificates_relation(event):
+            return
+
+        if not self._ensure_kube_control_relation(event):
+            return
+
+        channel = self.config.get("channel")
+        self._install_snaps(channel)
+
+        self.unit.status = ops.ActiveStatus("Ready to test.")
+
+    def _install_snaps(self, channel: Optional[str]) -> None:
+        self.unit.status = ops.MaintenanceStatus("Installing kubectl and kubernetes-test snaps.")
+        snap.ensure("kubectl", snap.SnapState.Latest, channel=channel)
+        snap.ensure("kubernetes-test", snap.SnapState.Latest, channel=channel, classic=True)
```

Comment:
this method needs the string values of those enums right?

```suggestion
        snap.ensure("kubectl", snap.SnapState.Latest.value, channel=channel)
        snap.ensure("kubernetes-test", snap.SnapState.Latest.value, channel=channel, classic=True)
```